lib/sles4sap/console_redirection.pm

```diff
@@ -0,0 +1,174 @@
+    $testapi::distri->{serial_term_prompt} = '';    # reset previous prompt first
+    my $current_user = script_output('whoami');
+    croak 'SSH user not defined and/or "whoami" script failed.' unless defined($current_user);
+    $testapi::distri->{serial_term_prompt} = ($current_user eq 'root' ? '# ' : '> ');
+}
+
+=head2 connect_target_to_serial
+
+    connect_target_to_serial([ssh_user=>ssh_user, target_ip=>$target_ip]);
+
+B<ssh_user> Login user - can be alternatively defined by OpenQA parameter REDIRECT_TARGET_USER
+B<target_ip> Target host IP - can be alternatively defined by OpenQA parameter REDIRECT_TARGET_IP
+Detects if login prompt appears and types the password.
+
+=cut
+
+sub connect_target_to_serial {
+    my (%args) = @_;
+    my $redirect_ip = $args{'target_ip'} // get_required_var('REDIRECT_TARGET_IP');
+    my $ssh_user = $args{'ssh_user'} // get_required_var('REDIRECT_TARGET_USER');
+
+    croak 'Missing "ssh_user" argument' unless $ssh_user;
+    croak 'Missing "redirect_ip" argument' unless $redirect_ip;
+    croak "OpenQA variable WORKER_VM_ID undefined. Run 'redirection_init()' first" unless get_var('WORKER_VM_ID');
+    croak "IP address '$redirect_ip' is not valid." unless verify_ip($redirect_ip);
+    croak 'Global variable "$searialdev" undefined' unless $serialdev;
+    croak "Console is already redirected to:" . script_output('hostname') if check_serial_redirection();
+
+    enter_cmd "ssh $ssh_opt $ssh_user\@$redirect_ip 2>&1 | tee -a /dev/$serialdev";
+    handle_login_prompt($ssh_user);
+    record_info('Redirect ON', "Serial redirection established to: $redirect_ip");
+}
+
+=head2 disconnect_target_from_serial
+
+    disconnect_target_from_serial($worker_id);
+
+Disconnects target from serial console by typing 'exit' command until host IP matches IP addr of the worker VM.
+
+=cut
+
+sub disconnect_target_from_serial {
+    my ($worker_id) = @_;
+    my $worker_machine_id = $worker_id // get_required_var('WORKER_VM_ID');
+    set_serial_term_prompt();
+    my $serial_redirection_status = check_serial_redirection($worker_machine_id);
+    while ($serial_redirection_status ne 0) {
+        enter_cmd('exit');    # Enter command and wait for screen start changing
+        $testapi::distri->{serial_term_prompt} = '';    # reset console prompt
+        wait_serial(qr/Connection.*closed./, timeout => 10);    # Wait for connection to close
+        wait_serial(qr/# |> /, timeout => 10);    # Wait for console prompt to appear
+        set_serial_term_prompt();    # after logout user might change and prompt with it.
+        $serial_redirection_status = check_serial_redirection($worker_machine_id);
+    }
+    record_info('Redirect OFF', "Serial redirection closed. Console set to: " . script_output('hostname'));
+}
+
+
+=head2 verify_ip
+
+    verify_ip($ip_address);
+
+B<ip_address> IP address
+
+Validates IP address format. Returns IP address if format is valied, 0 if invalid.
+
+=cut
+
+sub verify_ip {
```

Comment:
Does this strictly need to be exported? Can be a private internal method? Is it exported to be testable from UT, is it the only needs? Could be enough to directly call `grep(/^RE...` in the `connect_target_to_serial` ?